content/blog/22-23/reverse_engineering_epaper.md

```diff
@@ -0,0 +1,141 @@
+
+By reading the datasheet of the CC2510, it became clear that there is a debug interface. This interface can be used to put code on the board, to step through the code when debugging or even to read out the code. The debug interface consists of two signal pins: a host-to-microcontroller clock pin, and a bi-directional data pin. To enable the debug interface, the clock needs to be pulsed twice while the reset pin is held low. This makes for a total of 5 wires that need to be attached to the PCB: debug clock, debug data, reset; and then ground and 3.3V to power the board.
+
+<%= figure 'https://pics.zeus.gent/B84W2IoKSKqUvBO5bq3Hst0J1E8W1EtbJWgSgLsP.jpg', 'PCB mounted on 3D-printed holder, debug pins attached to Pi Pico. The battery is in the background, disconnected' %>
+
+The CC2510 microcontroller contains an 8051 processor core. 8051 is a rather old 8-bit instruction set, originally made by Intel, but it is used in a lot of embedded products. The debug interface is quite ingenious in how it implements most features: instead of individually imlementing write, read, verify and other features, it has the `DEBUG_INSTR` instruction. This debug instruction takes one to three bytes of arguments, and executes these as an 8051 instruction. After that, it sends the value of the accumulator (ACC/A) register back over the debug interface. Reading out the memory then is a loop of setting an address we want to read out and moving the value at that address into the ACC register.
+
+<%= figure 'https://pics.zeus.gent/Q0caDtNqNqD9D66klL8SFJQvfkTg014tvpVvJl7b.png', 'Table 45: Debug commands' %>
+
+The pins were attached to a Raspberry Pi Pico microcontroller, and a small application was written, based on [ESP_CC_Flasher](https://github.com/atc1441/ESP_CC_Flasher), to interface with debug interface of the microcontroller. Something unfortunate very quickly became clear: the chip was debug-locked: after programming the firmware on the microcontroller, the manufacturer disabled the debug interface. This means that only a very limited subset of debug instructions were enabled: only READ_STATUS (with which we can see, among others, if the chip is debug locked), GET_CHIP_ID (with which we can get the type of chip) and CHIP_ERASE (which erases the debug protection bits, but also all the other code). No instructions with which we can read out the board were found. Our new goal is thus to bypass the code read-out protection.
```

Comment:
```suggestion
The pins were attached to a Raspberry Pi Pico microcontroller, and a small application was written, based on [ESP_CC_Flasher](https://github.com/atc1441/ESP_CC_Flasher), to interface with debug interface of the microcontroller. Something unfortunate very quickly became clear: the chip was debug-locked: after programming the firmware on the microcontroller, the manufacturer disabled the debug interface. This means that only a very limited subset of debug instructions were enabled: only READ_STATUS (with which we can see, among others, if the chip is debug locked), GET_CHIP_ID (with which we can get the type of chip) and CHIP_ERASE (which erases the debug protection bits, but also all the other code). No instructions with which we can read out the board were found. Our new goal is thus to bypass the read-out protection of the code.
```

Reply:
dit is blijkbaar een courante term, dus ik ga dat houden